### PR TITLE
Logs: Add show context to dashboard panel

### PR DESCRIPTION
**Pull request by svennergr**
**What is this feature?**

This PR adds the "Log Context" functionality to the LogsPanel, i.e. it enables this feature for logs panels in dashboards.

**Special notes for your reviewer:**

- add a logs panel to a dashboard and find the show context toggle

### CODE REVIEW

**Comment by gtk-grafana** [comment]
Looking good! But it looks as though the log context button disappears when running multiple queries (even if hidden)?

image

image

image

**Comment by svennergr** [comment]
> Looking good! But it looks as though the log context button disappears when running multiple queries (even if hidden)?

Do you mind sharing steps to reproduce? I couldn't repro:

https://github.com/grafana/grafana/assets/8092184/8eceb39f-f1fe-4d7a-9fb1-142aab0034c3

File: public/app/plugins/panel/logs/LogsPanel.tsx [hunk 1]
```diff
@@ -37,14 +44,36 @@ export const LogsPanel = ({
...
   id,
 }: LogsPanelProps) => {
   const isAscending = sortOrder === LogsSortOrder.Ascending;
   const style = useStyles2(getStyles);
   const [scrollTop, setScrollTop] = useState(0);
   const logsContainerRef = useRef<HTMLDivElement>(null);
+  const [contextRow, setContextRow] = useState<LogRowModel | null>(null);
+  const [closeCallback, setCloseCallback] = useState<(() => void) | null>(null);
+
+  // create a map of refid -> datasource
```
Comment (matyax) [inline public/app/plugins/panel/logs/LogsPanel.tsx:58]: I don't know how you feel about this, but when components start growing in complexity and responsibilities it feels particularly uncomfortable, as it becomes challenging to maintain and test.

**Review by matyax — APPROVED** [review]
Works very well!

**Comment by gtk-grafana** [comment]
Also when you click into the logs context and you're using a mixed datasource it looks as though the query expression is including an empty vector selector, which throws an error in backend

image

image

**Comment by matyax** [comment]
Just submitted a review I started a few hours ago. I didn't see the rest of the activity in this PR, apologies in advance.

**Comment by matyax** [comment]
> the query expression is including an empty vector selector, which throws an error in backend

I was about to comment the same. I can intermitently reproduce it in main, so it seems unrelated with this PR. Looking into the root cause now.

https://github.com/grafana/grafana/assets/1069378/4254eaf3-74ea-4c8b-ba00-fba1c232ea68

**Comment by svennergr** [comment]
> > the query expression is including an empty vector selector, which throws an error in backend
> 
> I was about to comment the same. I can intermitently reproduce it in main, so it seems unrelated with this PR. Looking into the root cause now.
> 
>  Demo.mov

@matyax just fyi, as far as the Loki devenv, I've seen that failing due to Loki restarting a lot recently. I don't know whats wrong with the config in our devenv. Thanks for looking into that though.

**Comment by svennergr** [comment]
> Seems to only happen when the dashboard hasn't been saved yet

You mind retesting with the recent changes?

File: public/app/plugins/panel/logs/LogsPanel.tsx (hunk 1 above)
Comment (svennergr) [reply public/app/plugins/panel/logs/LogsPanel.tsx]: 100% agree. I extracted that logic into an additional hook. WDYT?

**Comment by gtk-grafana** [comment]
Nice! Appears to be working now

**Comment by gtk-grafana** [comment]
@matyax @svennergr I was able to repro that issue with mixed datasource using the Grafana Logs Dev and the SEDemo loki-logs as well as gdev-loki

File: public/app/plugins/panel/logs/LogsPanel.tsx [hunk 2]
```diff
@@ -102,28 +163,43 @@ export const LogsPanel = ({
...
-        {showCommonLabels && isAscending && renderCommonLabels()}
-      </div>
-    </CustomScrollbar>
+      )}
+      <CustomScrollbar autoHide scrollTop={scrollTop}>
+        <div className={style.container} ref={logsContainerRef}>
+          {showCommonLabels && !isAscending && renderCommonLabels()}
+          <LogRows
+            logRows={logRows}
+            // always return true for now as this propery will
```
Comment (gtk-grafana) [inline public/app/plugins/panel/logs/LogsPanel.tsx:182]: Is there more to this comment?

**Review by gtk-grafana — APPROVED** [review]
Looks good and is working for me locally. (since mixed issue appears to be in main, and not introduced by this PR)

File: public/app/plugins/panel/logs/LogsPanel.tsx (hunk 2 above)
Comment (svennergr) [reply public/app/plugins/panel/logs/LogsPanel.tsx:182]: Lol, removed that. Good catch!

**Pull request merged**